dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilder.java

```diff
@@ -356,19 +356,30 @@ private static void levelOrder(
   }
 
   /**
-   * Returns true if {@code exceptionAndPath.getPath} contains an {@code optional} dependency or a
+   * Records the path and {@code repositoryException} to {@code unsafeResolutionFailures} if the the
+   * exception at {@code failedDependencyNode} is not safe.
+   */
+  private static void recordUnsafeFailuresIfNotSafe(
+      Stack<DependencyNode> parentNodes,
+      DependencyNode failedDependencyNode,
+      RepositoryException repositoryException,
+      List<ExceptionAndPath> unsafeResolutionFailures) {
+    ExceptionAndPath failure =
```

Comment:
Would this be simpler?

```java
if (isOptionalOrProvided(parentNodes)) {
  unsafeResolutionFailures.add(
      ExceptionAndPath.create(parentNodes, collectionException.getResult().getRoot(), collectionException));
}
```

Reply:
`isOptionalOrProvided` also checks `failedDependencyNode`. The concatenation is done at `ExceptionAndPath create`.

If concatenation is done outside ExceptionAndPath, then that would be

```
if (isOptionalOrProvided(ImmutableList.<DependencyNode>builder()
                .addAll(parentDependencyNodes)
                .add(dependencyNode)
                .build())) {
  unsafeResolutionFailures.add(
      ExceptionAndPath.create(parentNodes, collectionException.getResult().getRoot(), collectionException));
}

```

So I think prev version is good.